front/tests/reporter/generate-personalized-report.ts

```diff
@@ -0,0 +1,278 @@
+    this.startTime = Date.now();
+    this.personalizedReport.results.summary.start = this.startTime;
+
+    const outputDir = this.reporterConfigOptions.outputDir!;
+    if (!fs.existsSync(outputDir)) {
+      fs.mkdirSync(outputDir, { recursive: true });
+    }
+
+    this.setOutputFilename(this.reporterConfigOptions.outputFile!);
+  }
+
+  /**
+   * Called at the end of the test run.
+   * Process all test results and write final report to disk.
+   */
+  onEnd() {
+    this.personalizedReport.results.summary.stop = Date.now();
+
+    if (this.suite && this.suite.allTests().length > 0) {
+      this.processSuite(this.suite);
+      this.personalizedReport.results.summary.suites = this.countSuites(this.suite);
+    }
+
+    this.writeReportToFile(this.personalizedReport);
+  }
+
+  /** Process test cases in the suite and nested child suites.
+   * Note: This value is not used yet, but will support future visualizations.
+   */
+  private processSuite(suite: Suite) {
+    suite.tests.forEach((test) => this.processTest(test));
+    suite.suites.forEach((child) => this.processSuite(child));
+  }
+
+  // Process a single test case by extracting its latest result.
+  private processTest(testCase: TestCase) {
+    const latestResult = testCase.results.at(-1);
+    if (!latestResult) return;
+
+    this.addTestResult(testCase, latestResult);
+    this.updateSummary(latestResult);
+  }
+
+  // Add a test case result to the report, with optional enrichment.
+  private addTestResult(testCase: TestCase, result: TestResult) {
+    const test: PersonalizedTest = {
+      name: testCase.title,
+      status: GenerateReport.mapStatus(result.status),
+      duration: result.duration,
+    };
+
+    if (!this.reporterConfigOptions.minimal) {
+      this.enrichTestData(test, testCase, result);
+    }
+
+    this.personalizedReport.results.tests.push(test);
+  }
+
+  // Add detailed metadata to the test result: timestamps, logs, retries, suite path, etc.
+  private enrichTestData(test: PersonalizedTest, testCase: TestCase, result: TestResult) {
+    test.start = GenerateReport.dateToUnix(result.startTime);
+    test.stop = GenerateReport.dateToUnix(result.startTime, result.duration);
+    test.rawStatus = result.status;
+    test.filePath = testCase.location.file;
+    test.retries = result.retry;
+    test.flaky = result.status === 'passed' && result.retry > 0;
+    test.steps = [];
+
+    const failure = GenerateReport.extractFailureDetails(result);
+    test.message = failure.message;
+    test.trace = failure.trace;
+
+    result.steps.forEach((step) => this.processStep(test, step));
+
+    test.suite = GenerateReport.buildSuitePath(testCase);
+
+    const metadata = GenerateReport.extractBrowserMetadata(result);
+    if (metadata?.name || metadata?.version) {
+      test.browser = `${metadata.name ?? ''} ${metadata.version ?? ''}`.trim();
+    }
+
+    test.stdout = result.stdout.map((item) =>
+      Buffer.isBuffer(item) ? item.toString() : String(item)
```

Comment:
Why do we need to special-case `Buffer` here? This is a `Buffer | string`, so `.toString()` should work on both.